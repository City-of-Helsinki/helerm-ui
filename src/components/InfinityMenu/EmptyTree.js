import React from 'react';

const EmptyTree = () => {
  return (
    <div className='alert alert-danger'>
      <p>Ei tuloksia</p>
    </div>
  );
};

export default EmptyTree;
