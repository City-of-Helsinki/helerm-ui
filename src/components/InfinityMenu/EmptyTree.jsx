import React from 'react';

const EmptyTree = () => (
  <div className='alert alert-danger'>
    <p>Ei tuloksia</p>
  </div>
);

export default EmptyTree;
