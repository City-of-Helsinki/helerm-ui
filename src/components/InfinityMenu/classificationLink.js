import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const stopPropagation = ev => ev.stopPropagation();

const ClassificationLink = ({ id }) => (
  <Link to={`/view-classification/${id}`} onClick={stopPropagation} title='Avaa luokituksen tiedot'>
    <i className='fa-solid fa-circle-info' />
  </Link>
);

ClassificationLink.propTypes = {
  id: PropTypes.string.isRequired
};

export default ClassificationLink;
