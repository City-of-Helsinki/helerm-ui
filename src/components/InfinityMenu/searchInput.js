import React, { PropTypes } from 'react';

const SearchInput = ({ setSearchInput, searchInput, placeholder, ...inputProps }) => (
  <input className='react-infinity-menu-default-search-input'
         {...inputProps}
         type='search'
         placeholder={placeholder}
         onChange={setSearchInput}
         value={searchInput}
  />
);

SearchInput.propTypes = {
  placeholder: PropTypes.string,
  searchInput: PropTypes.string,
  setSearchInput: PropTypes.func.isRequired
};

SearchInput.defaultProps = {
  placeholder: '🔍',
  searchInput: ''
};

export default SearchInput;
