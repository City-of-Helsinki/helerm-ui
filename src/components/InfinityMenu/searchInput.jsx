import React from 'react';
import PropTypes from 'prop-types';

const SearchInput = ({ setSearchInput, searchInput, placeholder, ...inputProps }) => (
  <input
    className='react-infinity-menu-default-search-input'
    {...inputProps}
    type='search'
    placeholder={placeholder}
    onChange={setSearchInput}
    value={searchInput}
  />
);

SearchInput.propTypes = {
  placeholder: PropTypes.string,
  searchInput: PropTypes.string,
  setSearchInput: PropTypes.func.isRequired,
};

SearchInput.defaultProps = {
  placeholder: '🔍',
  searchInput: '',
};

export default SearchInput;
