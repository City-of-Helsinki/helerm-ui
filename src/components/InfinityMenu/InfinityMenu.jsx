/* eslint-disable jsx-a11y/control-has-associated-label */
/* eslint-disable react/forbid-prop-types */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable no-param-reassign */
/* eslint-disable react/prop-types */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _get from 'lodash/get';
import cloneDeep from 'lodash/cloneDeep';
import classnames from 'classnames';
import { Link } from 'react-router-dom';
import NestedObjects from 'nested-objects';
import Defiant from 'defiant.js';
import Sticky from 'react-sticky-el';

import ClassificationLink from './ClassificationLink';
import EmptyTree from './EmptyTree';
import Exporter from '../Exporter';
import SearchInputs from './SearchInput/SearchInputs';
import SearchFilters from './SearchFilter/SearchFilters';

const DEFAULT_FILTER_CONDITION = 'and';

/**
 * Extracted from https://github.com/socialtables/react-infinity-menu
 */
class InfinityMenu extends Component {
  constructor(props) {
    super(props);
    this.state = {
      filteredTree: [],
      snapshots: {},
      filterCondition: DEFAULT_FILTER_CONDITION,
    };

    this.onNodeClick = this.onNodeClick.bind(this);
    this.onLoadMoreClick = this.onLoadMoreClick.bind(this);
    this.onFilterConditionChange = this.onFilterConditionChange.bind(this);
    this.filterTree = this.filterTree.bind(this);
    this.createSnapshots = this.createSnapshots.bind(this);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const { isDetailSearch, isSearchChanged, items, searchInputs, tree } = nextProps;

    const {
      isSearchChanged: wasSearchChanged,
      items: prevItems,
      searchInputs: prevInputs,
      tree: prevTree,
    } = this.props;

    if (isDetailSearch && items !== prevItems) {
      this.createSnapshots(items);
    }

    if (tree !== prevTree) {
      this.setState({ filteredTree: tree });
    }
    if (
      (tree !== prevTree && isSearchChanged && searchInputs) ||
      (isDetailSearch && isSearchChanged && !wasSearchChanged) ||
      (!isDetailSearch && searchInputs !== prevInputs)
    ) {
      this.filterTree(searchInputs, tree, isDetailSearch);
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (nextProps.shouldComponentUpdate) {
      return nextProps.shouldComponentUpdate(this.props, this.state, nextProps, nextState);
    }
    return true;
  }

  onNodeClick(tree, node, keyPath, event) {
    event.preventDefault();
    if (!this.props.isSearching) {
      node.isOpen = !node.isOpen;
      node.maxLeaves = this.props.maxLeaves;
      NestedObjects.set(tree, keyPath, node);
      if (this.props.onNodeMouseClick) {
        const currLevel = Math.floor(keyPath.split('.').length / 2);
        this.props.onNodeMouseClick(event, tree, node, currLevel, keyPath);
      }
    }
  }

  onLoadMoreClick(tree, node, keyPath, event) {
    event.preventDefault();
    // get parent node so we can increment it's unique max leaves property
    const keyPathArray = keyPath.split('.');
    const parentPath = Object.assign([], keyPathArray).splice(0, keyPathArray.length - 2);
    const parentNode = _get(this.props.tree, parentPath);
    // set new max leaves - if none exist use component default property
    parentNode.maxLeaves = !parentNode.maxLeaves ? this.props.maxLeaves : parentNode.maxLeaves + this.props.maxLeaves;
    if (this.props.onNodeMouseClick) {
      const currLevel = Math.floor(keyPath.split('.').length / 2);
      this.props.onNodeMouseClick(event, tree, node, currLevel, keyPath);
    }
  }

  onFilterConditionChange(option) {
    this.setState(
      (prevState) => ({ ...prevState, filterCondition: option.value }),
      () => this.filterTree(this.props.searchInputs, this.props.tree, this.props.isDetailSearch),
    );
  }

  getDetailFilters(searchInputs) {
    const filters = [];
    const { filterCondition } = this.state;
    searchInputs.forEach((input) => {
      const splitted = input.split('=').map((m) => m.trim());
      let filter;
      if (splitted.length === 2 && splitted[0] && splitted[1]) {
        filter = this.formatDetailFilter(splitted[0], splitted[1]);
      } else if (splitted[0]) {
        filter = this.formatDetailFilter('.', splitted[0]);
      }
      if (filter) {
        filters.push(filter);
      }
    });

    const filterConditionString = ` ${filterCondition} `;

    return filters.length ? `//*[${filters.join(filterConditionString)}]` : '';
  }

  // eslint-disable-next-line sonarjs/cognitive-complexity
  setDisplayTree(tree, prevs, curr, keyPath) {
    const currLevel = Math.floor(keyPath.length / 2);
    const currCustomComponent =
      typeof curr.customComponent === 'string'
        ? this.props.customComponentMappings[curr.customComponent]
        : curr.customComponent;
    const { isSearching } = this.props;
    const shouldDisplay = (isSearching && curr.isSearchDisplay) || !isSearching;
    curr.keyPath = keyPath;

    // the leaves
    if (!curr.children) {
      const keyPathArray = keyPath.split('.');
      const parentPath = Object.assign([], keyPathArray).splice(0, keyPathArray.length - 2);
      const parentNode = _get(tree, parentPath);
      const filteredChildren = parentNode.children.some((child) => child.isSearchDisplay === true)
        ? parentNode.children.filter((child) => child.isSearchDisplay === true)
        : parentNode.children;
      const itemKey = `infinity-menu-leaf-${curr.id}`;
      const visIds = filteredChildren.map((e) => e.id);

      let relativeIndex = visIds.indexOf(curr.id);
      relativeIndex = relativeIndex === -1 ? Infinity : relativeIndex;

      const parentMaxLeaves = parentNode.maxLeaves || this.props.maxLeaves;
      if (shouldDisplay && parentMaxLeaves > relativeIndex) {
        if (curr.customComponent) {
          const componentProps = {
            key: itemKey,
            onMouseDown: (e) => {
              if (this.props.onLeafMouseDown) {
                this.props.onLeafMouseDown(e, curr);
              }
            },
            onMouseUp: (e) => {
              if (this.props.onLeafMouseUp) {
                this.props.onLeafMouseUp(e, curr);
              }
            },
            onClick: (e) => {
              if (this.props.onLeafMouseClick) {
                this.props.onLeafMouseClick(e, curr);
              }
            },
            name: curr.name,
            icon: curr.icon,
            data: curr,
          };
          prevs.push(React.createElement(currCustomComponent, componentProps));
        } else {
          prevs.push(
            <li
              key={itemKey}
              className={`infinity-menu-leaf-container${!curr.function ? ' new-leaf' : ''}`}
              onMouseDown={(e) => (this.props.onLeafMouseDown ? this.props.onLeafMouseDown(e, curr) : null)}
              onMouseUp={(e) => (this.props.onLeafMouseUp ? this.props.onLeafMouseUp(e, curr) : null)}
              onClick={(e) => (this.props.onLeafMouseClick ? this.props.onLeafMouseClick(e, curr) : null)}
              onKeyUp={(e) => {
                if (this.props.onLeafMouseClick && e.key === 'Enter') {
                  this.props.onLeafMouseClick(e, curr);
                }
              }}
            >
              <span>
                {curr.name} <ClassificationLink id={curr.id} />
              </span>
            </li>,
          );
        }
      } else if (shouldDisplay && relativeIndex === filteredChildren.length - 1) {
        prevs.push(
          <li
            key={itemKey}
            className='infinity-menu-load-more-container'
            onClick={(e) => this.onLoadMoreClick(tree, curr, keyPath, e)}
            onKeyUp={(e) => {
              if (e.key === 'Enter') {
                this.onLoadMoreClick(tree, curr, keyPath, e);
              }
            }}
          >
            <span>N채yt채 lis채채</span>
          </li>,
        );
      }
      return prevs;
    }
    const key = `infinity-menu-node-${currLevel}-${curr.id}`;
    const nodeName = curr.name;
    if ((!curr.isOpen && !isSearching) || (!curr.isSearchOpen && isSearching)) {
      if (shouldDisplay) {
        if (curr.customComponent) {
          const nodeProps = {
            onClick: (e) => this.onNodeClick(tree, curr, keyPath, e),
            name: nodeName,
            isOpen: curr.isOpen,
            isSearching: false,
            data: curr,
            key,
          };
          prevs.push(React.createElement(currCustomComponent, nodeProps));
        } else {
          prevs.push(
            <button
              type='button'
              key={key}
              onClick={(e) => this.onNodeClick(tree, curr, keyPath, e)}
              onKeyUp={(e) => {
                if (e.key === 'Enter') {
                  this.onNodeClick(tree, curr, keyPath, e);
                }
              }}
              className={classnames('infinity-menu-node-container', {
                opened: !!curr.isOpen,
              })}
            >
              <label>
                {nodeName} <ClassificationLink id={curr.id} />
              </label>
            </button>,
          );
        }
      }
      return prevs;
    }
    const openedNode = [];
    if (shouldDisplay) {
      if (curr.customComponent) {
        const nodeProps = {
          onClick: (e) => this.onNodeClick(tree, curr, keyPath, e),
          name: nodeName,
          isOpen: curr.isOpen,
          data: curr,
          key,
          isSearching,
        };
        openedNode.push(React.createElement(currCustomComponent, nodeProps));
      } else {
        openedNode.push(
          <button
            type='button'
            key={key}
            onClick={(e) => this.onNodeClick(tree, curr, keyPath, e)}
            onKeyUp={(e) => {
              if (e.key === 'Enter') {
                this.onNodeClick(tree, curr, keyPath, e);
              }
            }}
            className={classnames('infinity-menu-node-container', {
              opened: !!curr.isOpen,
            })}
          >
            <label>
              {nodeName} <ClassificationLink id={curr.id} />
            </label>
          </button>,
        );
      }

      const childrenList = curr.children.length
        ? curr.children.reduce((p, c, k) => {
            if (c === undefined || k === undefined) {
              return p;
            }
            return this.setDisplayTree(tree, p, c, `${keyPath}.children.${k}`);
          }, [])
        : [];

      if (childrenList.length > 0) {
        openedNode.push(<ul key={`infinity-menu-children-list${currLevel}`}>{childrenList}</ul>);
      }
      prevs.push(openedNode);
    }
    return prevs;
  }

  getNodeMatchesSearchFilter(filters, node) {
    if (filters.length) {
      if (this.props.isDetailSearch) {
        const snapshot = this.state.snapshots[node.id] ? this.state.snapshots[node.id] : null;
        const result = snapshot ? JSON.search(snapshot, filters) : [];

        return result.length > 0;
      }

      return this.props.filter(node, filters);
    }

    return true;
  }

  filterTree(searchInputs, tree, isDetailSearch) {
    const filters = isDetailSearch ? this.getDetailFilters(searchInputs) : searchInputs[0];

    const filteredTree = filters
      ? tree.reduce((prev, curr, key) => {
          if (key === undefined) {
            return prev;
          }
          return this.findFiltered(prev, curr, key, filters);
        }, [])
      : tree;

    this.setState({ filteredTree });
  }

  createSnapshots(tree) {
    const snapshots = tree.reduce((prev, curr, key) => {
      if (key === undefined) {
        return prev;
      }
      return this.findSnapshot(prev, curr);
    }, {});
    this.setState({ snapshots });
  }

  findSnapshot(snapshots, node) {
    if (!node.children) {
      snapshots[node.id] = Defiant.getSnapshot(node);
      return snapshots;
    }
    return node.children.length
      ? node.children.reduce((prev, curr) => this.findSnapshot(prev, curr), snapshots)
      : snapshots;
  }

  findFiltered(trees, node, key, filters) {
    const newNode = cloneDeep(node);

    if (!node.children?.length) {
      const nodeMatchesSearchFilter = this.getNodeMatchesSearchFilter(filters, node);

      if (nodeMatchesSearchFilter) {
        newNode.isSearchDisplay = true;
        trees.push(newNode);
      }
      return trees;
    }

    const filteredSubFolder = node.children.length
      ? node.children.reduce((p, c, k) => this.findFiltered(p, c, k, filters), [])
      : [];

    const shouldDisplay = filteredSubFolder.some((child) => child.isSearchDisplay) || this.props.filter(node, filters);

    if (shouldDisplay) {
      newNode.isSearchOpen = true;
      newNode.children = filteredSubFolder;
      newNode.isSearchDisplay = true;
      newNode.maxLeaves = newNode.maxLeaves ? newNode.maxLeaves : this.props.maxLeaves;
      trees.push(newNode);
    }

    return trees;
  }

  // eslint-disable-next-line class-methods-use-this
  formatDetailFilter(field, value) {
    // returns xpath query e.g.
    // InformationSystem=*ahjo returns: contains(InformationSystem, "ahjo")
    // InformationSystem=ahjo* returns: starts-with(InformationSystem, "ahjo")
    // InformationSystem=ahjo returns: InformationSystem="ahjo"
    const wildcardIndex = value.indexOf('*');
    const fieldValue = value.replace(/\*/g, '');
    if (!fieldValue) {
      return null;
    }
    if (wildcardIndex === 0) {
      return `contains(${field}, "${fieldValue}")`;
    }
    if (wildcardIndex > 0) {
      return `starts-with(${field}, "${fieldValue}")`;
    }
    return `${field}="${fieldValue}"`;
  }

  renderBody(displayTree) {
    const { emptyTreeComponent, emptyTreeComponentProps, isFetching } = this.props;

    if (displayTree.length) {
      return displayTree;
    }
    if (emptyTreeComponent && !isFetching) {
      return React.createElement(emptyTreeComponent, emptyTreeComponentProps);
    }
    return null;
  }

  render() {
    const {
      attributeTypes,
      isDetailSearch,
      headerProps,
      searchInputs,
      addSearchInput,
      setSearchInput,
      removeSearchInput,
      isUser,
      filters,
      handleFilterChange,
    } = this.props;
    const { filteredTree, filterCondition } = this.state;

    // recursive go through the tree
    const displayTree = filteredTree.reduce((prev, curr, key) => {
      if (key === undefined) {
        return prev;
      }
      return this.setDisplayTree(filteredTree, prev, curr, key.toString());
    }, []);

    // header component
    const bodyContent = this.renderBody(displayTree);

    return (
      <div
        id='navigation-menu'
        className={classnames('navigation-menu', {
          'navigation-open': this.props.isOpen,
        })}
      >
        <Sticky stickyClassName='navigation-sticky'>
          <>
            <div className='navigation-header clearfix'>
              <button type='button' className='pull-left nav-button' onClick={this.props.toggleNavigationVisibility}>
                <span className='fa-solid fa-list' aria-hidden='true' />
              </button>
              {!!this.props.path.length && (
                <ol
                  className='breadcrumb'
                  onClick={this.props.toggleNavigationVisibility}
                  onKeyUp={(e) => {
                    if (e.key === 'Enter') {
                      this.props.toggleNavigationVisibility();
                    }
                  }}
                >
                  {this.props.path.map((item, index) => (
                    <li
                      className={classnames({
                        active: index === this.props.path.length,
                      })}
                      // eslint-disable-next-line react/no-array-index-key
                      key={index}
                    >
                      {item}
                    </li>
                  ))}
                </ol>
              )}
            </div>
            {this.props.isOpen && (
              <div className='navigation-container'>
                <div className='navigation-toggle'>
                  <button type='button' className='nav-button' onClick={this.props.toggleNavigationVisibility}>
                    <span className='fa-solid fa-xmark' aria-hidden='true' />
                  </button>
                </div>
                <div className='navigation-filters clearfix'>
                  <div className='navigation-filters-container'>
                    {isDetailSearch && (
                      <div className='row navigation-filters-header'>
                        <div className='col-xs-6'>
                          <h2>Sis채lt철haku</h2>
                        </div>
                        <div className='col-xs-6'>
                          <Exporter
                            attributeTypes={attributeTypes}
                            data={filteredTree}
                            isVisible={filteredTree.length > 0}
                          />
                        </div>
                      </div>
                    )}
                    <div className='row'>
                      <SearchInputs
                        headerProps={headerProps}
                        isDetailSearch={isDetailSearch}
                        searchInputs={searchInputs}
                        filterCondition={filterCondition}
                        addSearchInput={addSearchInput}
                        setSearchInput={setSearchInput}
                        removeSearchInput={removeSearchInput}
                        onFilterConditionChange={this.onFilterConditionChange}
                      />
                      <div
                        className={classnames({
                          'col-xs-12': isDetailSearch,
                          'col-sm-6': !isDetailSearch,
                        })}
                      >
                        <SearchFilters
                          attributeTypes={attributeTypes}
                          isDetailSearch={isDetailSearch}
                          isUser={isUser}
                          filters={filters}
                          handleFilterChange={handleFilterChange}
                        />
                      </div>
                    </div>
                  </div>
                  {!isDetailSearch && (
                    <div className='classification-link'>
                      <Link className='btn btn-primary btn-sm' to='/classification-tree'>
                        <span className='fa-solid fa-info' aria-hidden='true' />
                      </Link>
                    </div>
                  )}
                </div>
                <div className='infinity-menu-container'>
                  <div className='infinity-menu-display-tree-container'>{bodyContent}</div>
                </div>
              </div>
            )}
          </>
        </Sticky>
      </div>
    );
  }
}

InfinityMenu.propTypes = {
  addSearchInput: PropTypes.func.isRequired,
  attributeTypes: PropTypes.object,
  customComponentMappings: PropTypes.object,
  emptyTreeComponent: PropTypes.any,
  emptyTreeComponentProps: PropTypes.object,
  filter: PropTypes.func,
  headerProps: PropTypes.object,
  isDetailSearch: PropTypes.bool,
  isFetching: PropTypes.bool,
  isOpen: PropTypes.bool,
  isSearchChanged: PropTypes.bool,
  isSearching: PropTypes.bool,
  isUser: PropTypes.bool.isRequired,
  items: PropTypes.array,
  maxLeaves: PropTypes.number,
  onLeafMouseClick: PropTypes.func,
  onLeafMouseDown: PropTypes.func,
  onLeafMouseUp: PropTypes.func,
  onNodeMouseClick: PropTypes.func,
  path: PropTypes.array,
  removeSearchInput: PropTypes.func.isRequired,
  searchInputs: PropTypes.array.isRequired,
  setSearchInput: PropTypes.func.isRequired,
  toggleNavigationVisibility: PropTypes.func,
  tree: PropTypes.array,
  filters: PropTypes.object.isRequired,
  handleFilterChange: PropTypes.func.isRequired,
};

InfinityMenu.defaultProps = {
  emptyTreeComponent: EmptyTree,
  emptyTreeComponentProps: {},
  filter: (node, searchInput) => node.name.toLowerCase().indexOf(searchInput.toLowerCase()) >= 0,
  headerProps: {},
  maxLeaves: Infinity,
  onLeafMouseClick: () => {},
  onLeafMouseDown: () => {},
  onLeafMouseUp: () => {},
  onNodeMouseClick: () => {},
  tree: [],
};

export default InfinityMenu;
