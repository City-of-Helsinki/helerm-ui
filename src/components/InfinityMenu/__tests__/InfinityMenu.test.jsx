import { createBrowserHistory } from 'history';
import React from 'react';
import { Router } from 'react-router-dom';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import attributeRules from '../../../utils/mocks/attributeRules.json';
import validTOSwithChildren from '../../../utils/mocks/validTOSwithChildren.json';
import InfinityMenu from '../InfinityMenu';
import EmptyTree from '../EmptyTree';
import { navigationStateFilters, statusFilters } from '../../../constants';
import renderWithProviders from '../../../utils/renderWithProviders';

const baseMocks = {
  isDetailSearch: false,
  onNodeMouseClick: vi.fn(),
  filters: navigationStateFilters,
  filter: vi.fn(),
};

const renderComponent = (history, mocks = baseMocks) =>
  renderWithProviders(
    <Router history={history}>
      <InfinityMenu
        addSearchInput={vi.fn()}
        attributeTypes={attributeRules}
        customComponentMappings={{}}
        emptyTreeComponent={EmptyTree}
        emptyTreeComponentProps={{}}
        filter={mocks.filter}
        headerProps={{}}
        isDetailSearch={mocks.isDetailSearch}
        isFetching={false}
        isOpen
        isSearchChanged={false}
        isSearching={false}
        isUser
        items={[]}
        maxLeaves={Infinity}
        onLeafMouseClick={vi.fn()}
        onLeafMouseDown={vi.fn()}
        onLeafMouseUp={vi.fn()}
        onNodeMouseClick={mocks.onNodeMouseClick}
        path={[]}
        removeSearchInput={vi.fn()}
        searchInputs={['']}
        setSearchInput={vi.fn()}
        toggleNavigationVisibility={vi.fn()}
        tree={[]}
        filters={baseMocks.filters}
        handleFilterChange={vi.fn()}
      />
    </Router>,
    { history, preloadedState: { navigation: { items: [validTOSwithChildren] } } },
  );

describe('<InfinityMenu />', () => {
  it('should render correctly', () => {
    const history = createBrowserHistory();

    const { container } = renderComponent(history);

    expect(container).toMatchSnapshot();
  });

  it('should filter tree by searchInputs', async () => {
    const mockOnNodeMouseClick = vi.fn();
    const mockFilter = vi.fn();

    const mocks = { ...baseMocks, onNodeMouseClick: mockOnNodeMouseClick, filter: mockFilter };

    const history = createBrowserHistory();

    const { rerender } = renderComponent(history, mocks);

    rerender(
      <Router history={history}>
        <InfinityMenu
          addSearchInput={vi.fn()}
          attributeTypes={attributeRules}
          customComponentMappings={{}}
          emptyTreeComponent={EmptyTree}
          emptyTreeComponentProps={{}}
          filter={mocks.filter}
          headerProps={{}}
          isDetailSearch={mocks.isDetailSearch}
          isFetching={false}
          isOpen
          isSearchChanged
          isSearching={false}
          isUser
          items={[validTOSwithChildren]}
          maxLeaves={Infinity}
          onLeafMouseClick={vi.fn()}
          onLeafMouseDown={vi.fn()}
          onLeafMouseUp={vi.fn()}
          onNodeMouseClick={mocks.onNodeMouseClick}
          path={[]}
          removeSearchInput={vi.fn()}
          searchInputs={['Terveys']}
          setSearchInput={vi.fn()}
          toggleNavigationVisibility={vi.fn()}
          tree={[validTOSwithChildren]}
          filters={mocks.filters}
          handleFilterChange={vi.fn()}
        />
        ,
      </Router>,
    );

    expect(mockFilter).toHaveBeenCalled();
  });

  it('should filter tree by searchFilters', async () => {
    const mockOnNodeMouseClick = vi.fn();

    const mocks = {
      ...baseMocks,
      onNodeMouseClick: mockOnNodeMouseClick,
      filters: {
        ...baseMocks.filters,
        statusFilters: { ...baseMocks.filters.statusFilters, values: [statusFilters[3].value] },
      },
    };

    const history = createBrowserHistory();

    const { rerender } = renderComponent(history, mocks);

    rerender(
      <Router history={history}>
        <InfinityMenu
          addSearchInput={vi.fn()}
          attributeTypes={attributeRules}
          customComponentMappings={{}}
          emptyTreeComponent={EmptyTree}
          emptyTreeComponentProps={{}}
          filter={vi.fn()}
          headerProps={{}}
          isDetailSearch={mocks.isDetailSearch}
          isFetching={false}
          isOpen
          isSearchChanged={false}
          isSearching={false}
          isUser
          items={[validTOSwithChildren]}
          maxLeaves={Infinity}
          onLeafMouseClick={vi.fn()}
          onLeafMouseDown={vi.fn()}
          onLeafMouseUp={vi.fn()}
          onNodeMouseClick={mocks.onNodeMouseClick}
          path={[]}
          removeSearchInput={vi.fn()}
          searchInputs={['']}
          setSearchInput={vi.fn()}
          toggleNavigationVisibility={vi.fn()}
          tree={[validTOSwithChildren]}
          filters={mocks.filters}
          handleFilterChange={vi.fn()}
        />
        ,
      </Router>,
    );

    const infinityMenuItem = screen.getByRole('button', { name: '00 Hallintoasiat' });

    expect(infinityMenuItem).toBeInTheDocument();

    const user = userEvent.setup();

    await user.click(infinityMenuItem);

    expect(mockOnNodeMouseClick).toHaveBeenCalled();
  });
});
