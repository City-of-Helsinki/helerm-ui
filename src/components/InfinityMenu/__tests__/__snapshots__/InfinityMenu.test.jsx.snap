// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<InfinityMenu /> > should render correctly 1`] = `
<div>
  <div
    class="navigation-menu navigation-open"
    id="navigation-menu"
  >
    <div
      style="min-height: 0px;"
    >
      <div
        class=" "
        style="transform: translateZ(0);"
      >
        <div
          class="navigation-header clearfix"
        >
          <button
            class="pull-left nav-button"
            type="button"
          >
            <span
              aria-hidden="true"
              class="fa-solid fa-list"
            />
          </button>
        </div>
        <div
          class="navigation-container"
        >
          <div
            class="navigation-toggle"
          >
            <button
              class="nav-button"
              type="button"
            >
              <span
                aria-hidden="true"
                class="fa-solid fa-xmark"
              />
            </button>
          </div>
          <div
            class="navigation-filters clearfix"
          >
            <div
              class="navigation-filters-container"
            >
              <div
                class="row"
              >
                <div
                  class="col-sm-6"
                >
                  <input
                    class="react-infinity-menu-default-search-input"
                    placeholder="Etsi..."
                    type="search"
                    value=""
                  />
                </div>
                <div
                  class="col-sm-6"
                >
                  <div
                    class=""
                  >
                    <div
                      class=""
                    >
                      <div
                        class="Select css-2b097c-container"
                      >
                        <div
                          class=" css-yk16xz-control"
                        >
                          <div
                            class=" css-g1d714-ValueContainer"
                          >
                            <div
                              class=" css-1wa3eu0-placeholder"
                            >
                              Suodata viimeisen tilan mukaan...
                            </div>
                            <div
                              class="css-b8ldur-Input"
                            >
                              <div
                                class=""
                                style="display: inline-block;"
                              >
                                <input
                                  aria-autocomplete="list"
                                  autocapitalize="none"
                                  autocomplete="off"
                                  autocorrect="off"
                                  id="react-select-2-input"
                                  spellcheck="false"
                                  style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                                  tabindex="0"
                                  type="text"
                                  value=""
                                />
                                <div
                                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                                />
                              </div>
                            </div>
                          </div>
                          <div
                            class=" css-1hb7zxy-IndicatorsContainer"
                          >
                            <span
                              class=" css-1okebmr-indicatorSeparator"
                            />
                            <div
                              aria-hidden="true"
                              class=" css-tlfecz-indicatorContainer"
                            >
                              <svg
                                aria-hidden="true"
                                class="css-6q0nyr-Svg"
                                focusable="false"
                                height="20"
                                viewBox="0 0 20 20"
                                width="20"
                              >
                                <path
                                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                                />
                              </svg>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="classification-link"
            >
              <a
                class="btn btn-primary btn-sm"
                href="/classification-tree"
              >
                <span
                  aria-hidden="true"
                  class="fa-solid fa-info"
                />
              </a>
            </div>
          </div>
          <div
            class="infinity-menu-container"
          >
            <div
              class="infinity-menu-display-tree-container"
            >
              <div
                class="alert alert-danger"
              >
                <p>
                  Ei tuloksia
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
